In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

from sklearn.ensemble import RandomForestClassifier

In [2]:
raw_data = pd.read_csv('1.04. Real-life example.csv')

In [3]:
raw_data.head()

Brand    Price       Body  Mileage  EngineV Engine Type  \
0            BMW   4200.0      sedan      277      2.0      Petrol   
1  Mercedes-Benz   7900.0        van      427      2.9      Diesel   
2  Mercedes-Benz  13300.0      sedan      358      5.0         Gas   
3           Audi  23000.0  crossover      240      4.2      Petrol   
4         Toyota  18300.0  crossover      120      2.0      Petrol   

  Registration  Year         Model  
0          yes  1991           320  
1          yes  1999  Sprinter 212  
2          yes  2003         S 500  
3          yes  2007            Q7  
4          yes  2011         Rav 4

In [4]:
raw_data.describe(include = 'all')

Brand          Price   Body      Mileage      EngineV  \
count         4345    4173.000000   4345  4345.000000  4195.000000   
unique           7            NaN      6          NaN          NaN   
top     Volkswagen            NaN  sedan          NaN          NaN   
freq           936            NaN   1649          NaN          NaN   
mean           NaN   19418.746935    NaN   161.237284     2.790734   
std            NaN   25584.242620    NaN   105.705797     5.066437   
min            NaN     600.000000    NaN     0.000000     0.600000   
25%            NaN    6999.000000    NaN    86.000000     1.800000   
50%            NaN   11500.000000    NaN   155.000000     2.200000   
75%            NaN   21700.000000    NaN   230.000000     3.000000   
max            NaN  300000.000000    NaN   980.000000    99.990000   

       Engine Type Registration         Year    Model  
count         4345         4345  4345.000000     4345  
unique           4            2          NaN      312  
top         Diesel          yes          NaN  E-Class  
freq          2019         3947          NaN      199  
mean           NaN          NaN  2006.550058      NaN  
std            NaN          NaN     6.719097      NaN  
min            NaN          NaN  1969.000000      NaN  
25%            NaN          NaN  2003.000000      NaN  
50%            NaN          NaN  2008.000000      NaN  
75%            NaN          NaN  2012.000000      NaN  
max            NaN          NaN  2016.000000      NaN

In [5]:
raw_data.isnull().sum()

Brand             0
Price           172
Body              0
Mileage           0
EngineV         150
Engine Type       0
Registration      0
Year              0
Model             0
dtype: int64

In [6]:
raw_data = raw_data.dropna(subset=['Price'])

In [7]:
raw_data = raw_data.dropna(subset=['EngineV'])

In [8]:
raw_data.isnull().sum()

Brand           0
Price           0
Body            0
Mileage         0
EngineV         0
Engine Type     0
Registration    0
Year            0
Model           0
dtype: int64

In [9]:
raw_data.columns

Index(['Brand', 'Price', 'Body', 'Mileage', 'EngineV', 'Engine Type',
       'Registration', 'Year', 'Model'],
      dtype='object')

DROPPING OUTLIERS

In [10]:
raw_data.drop(raw_data[raw_data ['EngineV'] > 6.5].index,  inplace=True)


In [11]:
raw_data.drop(raw_data[raw_data ['Price'] > 100000].index,  inplace=True)

In [12]:
raw_data.drop(raw_data[raw_data ['Year'] < 1998].index,  inplace=True)

In [13]:
raw_data

Brand    Price       Body  Mileage  EngineV Engine Type  \
1     Mercedes-Benz   7900.0        van      427      2.9      Diesel   
2     Mercedes-Benz  13300.0      sedan      358      5.0         Gas   
3              Audi  23000.0  crossover      240      4.2      Petrol   
4            Toyota  18300.0  crossover      120      2.0      Petrol   
7              Audi  14200.0      vagon      200      2.7      Diesel   
...             ...      ...        ...      ...      ...         ...   
4337        Renault   6800.0      sedan      152      1.6      Petrol   
4338     Volkswagen  11500.0        van      163      2.5      Diesel   
4339         Toyota  17900.0      sedan       35      1.6      Petrol   
4341            BMW   6500.0      sedan        1      3.5      Petrol   
4344     Volkswagen  13500.0        van      124      2.0      Diesel   

     Registration  Year              Model  
1             yes  1999       Sprinter 212  
2             yes  2003              S 500  
3             yes  2007                 Q7  
4             yes  2011              Rav 4  
7             yes  2006                 A6  
...           ...   ...                ...  
4337          yes  2007             Megane  
4338          yes  2008   T5 (Transporter)  
4339          yes  2014            Corolla  
4341          yes  1999                535  
4344          yes  2013  T5 (Transporter)   

[3504 rows x 9 columns]

In [14]:
raw_data.describe(include = 'all')

Brand          Price   Body      Mileage      EngineV  \
count         3504    3504.000000   3504  3504.000000  3504.000000   
unique           7            NaN      6          NaN          NaN   
top     Volkswagen            NaN  sedan          NaN          NaN   
freq           777            NaN   1240          NaN          NaN   
mean           NaN   18310.428613    NaN   154.236587     2.455123   
std            NaN   16467.265205    NaN    92.748360     0.966150   
min            NaN    1200.000000    NaN     0.000000     0.600000   
25%            NaN    8100.000000    NaN    90.000000     1.800000   
50%            NaN   12500.000000    NaN   150.000000     2.200000   
75%            NaN   22500.000000    NaN   212.000000     3.000000   
max            NaN  100000.000000    NaN   980.000000     6.500000   

       Engine Type Registration         Year Model  
count         3504         3504  3504.000000  3504  
unique           4            2          NaN   267  
top         Diesel          yes          NaN  Vito  
freq          1701         3190          NaN   153  
mean           NaN          NaN  2007.825342   NaN  
std            NaN          NaN     4.713096   NaN  
min            NaN          NaN  1998.000000   NaN  
25%            NaN          NaN  2004.000000   NaN  
50%            NaN          NaN  2008.000000   NaN  
75%            NaN          NaN  2012.000000   NaN  
max            NaN          NaN  2016.000000   NaN

dropping model because there are many models (267) to convert in a dummy variable

In [15]:
raw_data = raw_data.drop(['Model'], axis = 1)

In [16]:
raw_data.describe(include = 'all')

Brand          Price   Body      Mileage      EngineV  \
count         3504    3504.000000   3504  3504.000000  3504.000000   
unique           7            NaN      6          NaN          NaN   
top     Volkswagen            NaN  sedan          NaN          NaN   
freq           777            NaN   1240          NaN          NaN   
mean           NaN   18310.428613    NaN   154.236587     2.455123   
std            NaN   16467.265205    NaN    92.748360     0.966150   
min            NaN    1200.000000    NaN     0.000000     0.600000   
25%            NaN    8100.000000    NaN    90.000000     1.800000   
50%            NaN   12500.000000    NaN   150.000000     2.200000   
75%            NaN   22500.000000    NaN   212.000000     3.000000   
max            NaN  100000.000000    NaN   980.000000     6.500000   

       Engine Type Registration         Year  
count         3504         3504  3504.000000  
unique           4            2          NaN  
top         Diesel          yes          NaN  
freq          1701         3190          NaN  
mean           NaN          NaN  2007.825342  
std            NaN          NaN     4.713096  
min            NaN          NaN  1998.000000  
25%            NaN          NaN  2004.000000  
50%            NaN          NaN  2008.000000  
75%            NaN          NaN  2012.000000  
max            NaN          NaN  2016.000000

BECAUSE THE MODEL IS RANDOM FOREST WE NEED TO CATEGORIZE PRICE

In [17]:
bins = [0, 1000, 5000, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000, 900000]
names = ['<1000', '1k-5k', '5k-10k', '10k-20k', '20k-30k','30k-40k', '40k-50k','50k-60k','60k-70k','70k-80k','80k-90k''90k+']

In [18]:
d = dict(enumerate(names, 1))

In [19]:
raw_data['Price Range'] = np.vectorize(d.get)(np.digitize(raw_data['Price'], bins))

In [20]:
raw_data = raw_data.drop(['Price'], axis = 1)

In [21]:
raw_data.head()

Brand       Body  Mileage  EngineV Engine Type Registration  Year  \
1  Mercedes-Benz        van      427      2.9      Diesel          yes  1999   
2  Mercedes-Benz      sedan      358      5.0         Gas          yes  2003   
3           Audi  crossover      240      4.2      Petrol          yes  2007   
4         Toyota  crossover      120      2.0      Petrol          yes  2011   
7           Audi      vagon      200      2.7      Diesel          yes  2006   

  Price Range  
1      5k-10k  
2     10k-20k  
3     20k-30k  
4     10k-20k  
7     10k-20k

In [22]:
raw_data.columns

Index(['Brand', 'Body', 'Mileage', 'EngineV', 'Engine Type', 'Registration',
       'Year', 'Price Range'],
      dtype='object')

In [23]:
raw_data = raw_data[['Price Range', 'Brand', 'Body', 'Mileage', 'EngineV', 'Engine Type', 'Registration',
       'Year']]

In [24]:
raw_data.head()

Price Range          Brand       Body  Mileage  EngineV Engine Type  \
1      5k-10k  Mercedes-Benz        van      427      2.9      Diesel   
2     10k-20k  Mercedes-Benz      sedan      358      5.0         Gas   
3     20k-30k           Audi  crossover      240      4.2      Petrol   
4     10k-20k         Toyota  crossover      120      2.0      Petrol   
7     10k-20k           Audi      vagon      200      2.7      Diesel   

  Registration  Year  
1          yes  1999  
2          yes  2003  
3          yes  2007  
4          yes  2011  
7          yes  2006

DUMMY VARIABLES

In [25]:
from sklearn.preprocessing import LabelEncoder

In [26]:
le_Brand = LabelEncoder()
le_Body = LabelEncoder()
le_Engine_Type = LabelEncoder()
le_Registration = LabelEncoder()
le_Price_Range = LabelEncoder()

In [27]:
raw_data['Brand_n'] = le_Brand.fit_transform(raw_data['Brand'])
raw_data['Body_n'] = le_Body.fit_transform(raw_data['Body'])
raw_data['Engine_Type_n'] = le_Engine_Type.fit_transform(raw_data['Engine Type'])
raw_data['Registration_n'] = le_Registration.fit_transform(raw_data['Registration'])
raw_data['Price_Range_n'] = le_Price_Range.fit_transform(raw_data['Price Range'])

In [28]:
raw_data.head()

Price Range          Brand       Body  Mileage  EngineV Engine Type  \
1      5k-10k  Mercedes-Benz        van      427      2.9      Diesel   
2     10k-20k  Mercedes-Benz      sedan      358      5.0         Gas   
3     20k-30k           Audi  crossover      240      4.2      Petrol   
4     10k-20k         Toyota  crossover      120      2.0      Petrol   
7     10k-20k           Audi      vagon      200      2.7      Diesel   

  Registration  Year  Brand_n  Body_n  Engine_Type_n  Registration_n  \
1          yes  1999        2       5              0               1   
2          yes  2003        2       3              1               1   
3          yes  2007        0       0              3               1   
4          yes  2011        5       0              3               1   
7          yes  2006        0       4              0               1   

   Price_Range_n  
1              6  
2              0  
3              2  
4              0  
7              0

In [29]:
raw_data = raw_data.drop(['Brand', 'Body', 'Engine Type', 'Registration', 'Price Range'], axis = 1)

In [30]:
raw_data.head() 

Mileage  EngineV  Year  Brand_n  Body_n  Engine_Type_n  Registration_n  \
1      427      2.9  1999        2       5              0               1   
2      358      5.0  2003        2       3              1               1   
3      240      4.2  2007        0       0              3               1   
4      120      2.0  2011        5       0              3               1   
7      200      2.7  2006        0       4              0               1   

   Price_Range_n  
1              6  
2              0  
3              2  
4              0  
7              0

In [31]:
raw_data.reset_index(drop = True)

Mileage  EngineV  Year  Brand_n  Body_n  Engine_Type_n  Registration_n  \
0         427      2.9  1999        2       5              0               1   
1         358      5.0  2003        2       3              1               1   
2         240      4.2  2007        0       0              3               1   
3         120      2.0  2011        5       0              3               1   
4         200      2.7  2006        0       4              0               1   
...       ...      ...   ...      ...     ...            ...             ...   
3499      152      1.6  2007        4       3              3               1   
3500      163      2.5  2008        6       5              0               1   
3501       35      1.6  2014        5       3              3               1   
3502        1      3.5  1999        1       3              3               1   
3503      124      2.0  2013        6       5              0               1   

      Price_Range_n  
0                 6  
1                 0  
2                 2  
3                 0  
4                 0  
...             ...  
3499              6  
3500              0  
3501              0  
3502              6  
3503              0  

[3504 rows x 8 columns]

In [32]:
data_clean = raw_data.copy()

In [33]:
data_clean.head()

Mileage  EngineV  Year  Brand_n  Body_n  Engine_Type_n  Registration_n  \
1      427      2.9  1999        2       5              0               1   
2      358      5.0  2003        2       3              1               1   
3      240      4.2  2007        0       0              3               1   
4      120      2.0  2011        5       0              3               1   
7      200      2.7  2006        0       4              0               1   

   Price_Range_n  
1              6  
2              0  
3              2  
4              0  
7              0

In [34]:
data_clean.describe(include = 'all')

Mileage      EngineV         Year      Brand_n       Body_n  \
count  3504.000000  3504.000000  3504.000000  3504.000000  3504.000000   
mean    154.236587     2.455123  2007.825342     3.257991     2.557078   
std      92.748360     0.966150     4.713096     2.078723     1.698061   
min       0.000000     0.600000  1998.000000     0.000000     0.000000   
25%      90.000000     1.800000  2004.000000     1.000000     1.000000   
50%     150.000000     2.200000  2008.000000     3.000000     3.000000   
75%     212.000000     3.000000  2012.000000     5.000000     4.000000   
max     980.000000     6.500000  2016.000000     6.000000     5.000000   

       Engine_Type_n  Registration_n  Price_Range_n  
count    3504.000000     3504.000000    3504.000000  
mean        1.243151        0.910388       2.809075  
std         1.363112        0.285666       2.657286  
min         0.000000        0.000000       0.000000  
25%         0.000000        1.000000       0.000000  
50%         1.000000        1.000000       2.000000  
75%         3.000000        1.000000       6.000000  
max         3.000000        1.000000       9.000000

SELECTING INPUTS AND TARGET

In [35]:
x = data_clean.drop(['Price_Range_n'], axis = 1)

In [36]:
y = data_clean['Price_Range_n']

In [37]:
y.head()

1    6
2    0
3    2
4    0
7    0
Name: Price_Range_n, dtype: int32

SELECTING DATA TO BE TESTED AND TRAINING

In [38]:
from sklearn.model_selection import train_test_split

In [39]:
x_train, x_test, y_train,  y_test = train_test_split(x,y, test_size = 0.2, random_state=42)

NOW RUNNING THE MODEL RANDOM FOREST

In [40]:
model = RandomForestClassifier()

In [41]:
model.fit(x_train,y_train )

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [45]:
p = model.predict(x_test)

In [43]:
model.score(x_test, y_test)

0.6990014265335235

In [46]:
compare = pd.DataFrame()
compare['y_test'] = y_test
compare['predicting'] = p
compare['Matching'] = np.where(y_test == p, 1, 0)
compare.describe()

y_test  predicting    Matching
count  701.000000  701.000000  701.000000
mean     2.736091    2.808845    0.699001
std      2.651732    2.651788    0.459020
min      0.000000    0.000000    0.000000
25%      0.000000    0.000000    0.000000
50%      2.000000    2.000000    1.000000
75%      6.000000    6.000000    1.000000
max      9.000000    9.000000    1.000000

In [47]:
counts = (compare['Matching']==1).value_counts().to_dict()
counts

{True: 490, False: 211}

In [48]:
accuracy = counts[True]/(counts[True] + counts[False])

In [49]:
accuracy

0.6990014265335235

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
cm = confusion_matrix(p, y_test)
cm

In [ ]:
#MAKING A DATAFRAME FOR THE CONFUSION MATRIX

cm_df = pd.DataFrame(cm)
cm_df.columns = ['P 0','P 1','P 2','P 3','P 4','P 5','P 6','P 7','P 8','P 9']
cm_df = cm_df.rename(index={0: 'Actual 0',1:'Actual 1',2:'Actual 2',3:'Actual 3',4:'Actual 4',5:'Actual 5',6:'Actual 6',7:'Actual 7',8:'Actual 8',9:'Actual 9'})
cm_df